test/Microsoft.ML.TestFramework/RemoteExecutorTestBase.cs

```diff
@@ -0,0 +1,243 @@
+            }
+
+            private void Dispose(bool disposing)
+            {
+                Assert.True(disposing, $"A test {AssemblyName}!{ClassName}.{MethodName} forgot to Dispose() the result of RemoteInvoke()");
+
+                if (Process != null)
+                {
+                    // A bit unorthodox to do throwing operations in a Dispose, but by doing it here we avoid
+                    // needing to do this in every derived test and keep each test much simpler.
+                    try
+                    {
+                        Assert.True(Process.WaitForExit(Options.TimeOut),
+                            $"Timed out after {Options.TimeOut}ms waiting for remote process {Process.Id}");
+
+                        if (File.Exists(Options.ExceptionFile))
+                        {
+                            throw new RemoteExecutionException(File.ReadAllText(Options.ExceptionFile));
+                        }
+
+                        if (Options.CheckExitCode)
+                        {
+                            int expected = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? Options.ExpectedExitCode : unchecked((sbyte)Options.ExpectedExitCode);
+                            int actual = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? Process.ExitCode : unchecked((sbyte)Process.ExitCode);
+
+                            Assert.True(expected == actual, $"Exit code was {Process.ExitCode} but it should have been {Options.ExpectedExitCode}");
+                        }
+                    }
+                    finally
+                    {
+                        if (File.Exists(Options.ExceptionFile))
+                        {
+                            File.Delete(Options.ExceptionFile);
+                        }
+
+                        // Cleanup
+                        try { Process.Kill(); }
+                        catch { } // ignore all cleanup errors
+
+                        Process.Dispose();
+                        Process = null;
+                    }
+                }
+            }
+
+            ~RemoteInvokeHandle()
+            {
+                // Finalizer flags tests that omitted the explicit Dispose() call; they must have it, or they aren't
+                // waiting on the remote execution
+                Dispose(disposing: false);
+            }
+
+            private sealed class RemoteExecutionException : XunitException
+            {
+                internal RemoteExecutionException(string stackTrace) : base("Remote process failed with an unhandled exception.", stackTrace) { }
+            }
+        }
+    }
+
+    /// <summary>Options used with RemoteInvoke.</summary>
+    public sealed class RemoteInvokeOptions
+    {
+        public bool Start { get; set; } = true;
+        public ProcessStartInfo StartInfo { get; set; } = new ProcessStartInfo();
+        public bool CheckExitCode { get; set; } = true;
+        public string mode { get; set; }
```

Comment:
`mode` => `Mode`

Reply:
Actually, we should probably make this a little more general purpose - as it stands it is really specific to HW intrinsics.

Maybe instead of using `mode`, the should be setting the EnvironmentVariables on the StartInfo.

---
In reply to: [249941399](https://github.com/dotnet/machinelearning/pull/2200#discussion_r249941399) [](ancestors = 249941399)